design/DFX-Interface.md

```diff
@@ -99,22 +91,27 @@ This file informs Motoko about the interface of that canister. It could be the o
 
 Open problem: how to resolve mutual canister imports.
 
-Compiling IDL Files to JS
--------------------------
-
-In order to compile an IDL file, `dfx` invokes `didc` with
-
-    didc --js some/path/input.did -o another/path/output.js
-
-This _reads_ `some/path/input.did` and any `.did` file referenced by
-`some/path/input.did`.
-
-No constraints are imposed where these imported files reside (this may be refined to prevent relative imports from looking outside the project and the declared packages)
-
-This _writes_ to `another/path/output.js`, but has no other effect. It does
-not create `another/path/`.
-
-Compiler warnings and errors are reported to `stderr`. Nothing writes to `stdout`.
+Exporting Canister Metadata
+---------------------------
+
+The compiler generates various metadata about the canister via command line flags.
+The compiled Wasm module also includes the metadata in the custom sections.
+
+* Candid interface.
+  + Compiler flag `--idl` generates the Candid interface for the canister. The main service
+    is always a service constructor, which contains the initialization arguments for installing the canister.
+  + Custom section `icp:public candid` stores the interface for the running (initialized) canister, which removes
+    the initializatio arguments. This custom section is publicly accessible.
+  + Custom section `icp:private init_args` stores the initialization arguments. The argument types can refer to
```

Comment:
I suppose we want to have first class service constructors eventually, in which case this would have to be public.

I would suggest to make the public/private choice independent from the section type. For example, a dev might install a private canister that they do not want to publish the interface for. But they'll still need the candid section privately to verify upgrades.

Reply:
But how will they access that section? By inspecting the wasm or querying the IC. I think I need to reread the IC spec but were you not advocating that the IC only serves custom sections named "public" or is there now support for controllers to access "private" sections?